video: introduction of STM32 DCMIPP driver

avolmat-st opened the pull request:
This PR is the introduction of the DCMIPP (camera pipeline) for STM32. It currently supports the STM32N6, however it is also applicable on some other SOCs such as STM32H7RS or STM32MP13xx. (for those latter ones, some modifications are necessary since they have simpler pipelines and no CSI hence new compatible will be needed).

This PR depends on the PR fixing the 8 bit raw bayer formats: #89397

Some TODOs / FIXME remains within the driver which I plan to correct prior to merging this PR however I first publish the PR to hopefully get some first comments.

Some points might benefit to other drivers as well and as such would be candidate to go to the video-common part. For example I've added a macro to get the COLORSPACE from a pixelformat in order to know if it is YUV / RGB or Bayer. Moreover, as far as I know Zephyr video format currently do not have yet information about quantization which would be necessary in order to know if a format is encoded on the limited or full range. Indeed, the DCMIPP performs color-conversion and conversion matrix depends on the range (for now I've only kept simple color conversion while waiting for quantization).

I also chose to add Kconfig entries in otder to define what is expected to be use for the sensor format. Indeed, since the DCMIPP can perform format conversion, I feel that it makes the driver (hence code size) much simpler to be able to define which format / size is to be used from the sensor rather than having to figure out that within the driver. Here as well, some other camera pipeline might benefit from it, in which case thise could go to the common Kconfig.

Commit:
dts-bindings: video: addition of stm32 dcmipp description

Addition of description for the STM32
Digital Camera Memory Interface Pixel Processor (DCMIPP).

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dts: arm: st: n6: add dcmipp node

Add node describing the dcmipp in stm32n6.dtsi

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
boards: st: stm32n6570_dk: csi_22pins connector related labels

Add csi_22pins_connector and related label in order to use
shields relying on csi_pins_connector.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
boards: st: stm32n6570_dk: Add IC17/IC18 settings for DCMIPP/CSI

Add configuration of the IC17 and IC18 clock dividers, used
by the dcmipp and csi IPs in the stm32n6.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
drivers: video: introduction of the stm32 DCMIPP driver

The STM32 Digital Camera Memory Interface Pixel Processor (DCMIPP)
is a multi-pipeline camera interface allowing to capture
and process frames from parallel or CSI interfaces depending on its
version.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

erwango commented on drivers/video/video_stm32_dcmipp.c:
See https://github.com/zephyrproject-rtos/zephyr/pull/85564 for equivalent functionality

Commit:
dts: arm: st: mp13: add dcmipp node in stm32mp135.dtsi

Add node describing the DCMIPP available on stm32mp135.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

josuah reviewed (commented):
I will have to come back to it as ran out of time but hopefully this can help in the meantime!
Looking forward to try this DK!

josuah reviewed (changes requested):
Actually here is one thing you might be interested in:

- Adding the board to the "build_all" video driver test, so that CI always checks that it at least build when someone else changes i.e. some API.

-> https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/build_all/video/boards
-> https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/build_all/video/testcase.yaml

[EDIT: I flagged as "change requested" just to raise awareness, but it is not up to me to decide I think]

ngphibang commented on dts/bindings/video/st,stm32-dcmipp.yaml:
```diff
@@ -0,0 +1,55 @@
+  resets:
+    required: true
+
+  sensor:
+    required: true
+    type: phandle
+    description: phandle of connected sensor device
```
This is not needed as the sensor can be retrieved from the port / endpoint (similar to Linux way).

ngphibang commented on drivers/video/video_stm32_dcmipp.c:
You can use the port / endpoint macros to retrieve the sensor dev

avolmat-st commented:
Thank you for all those comments. I am almost done with the addition of the parallel interface addition so I'll push an update of the driver with parallel interface / possibility to instanciate it on other SOCs (with single pipe) and take care of the comments you provided. Thanks again.

josuah commented on boards/st/stm32n6570_dk/stm32n6570_dk_common.dtsi:
```diff
@@ -46,6 +46,21 @@
 		};
 	};
+
+	csi_camera_connector: connector_csi_camera {
+		compatible = "st,stm32-csi-camera";
```
It seems like [this camera module](https://www.st.com/en/evaluation-tools/b-cams-imx.html?ecmp=tt9470_gl_link_feb2019&rt=db&id=DB5351) is using the de-facto standard 22-pin connector introduced by Raspberry Pi.

It might be possible to define an extra alias label i.e. at the bottom of the `.dts`:
https://github.com/zephyrproject-rtos/zephyr/blob/4a31b8036d9c20349a0ee9ab83fa86147fdf983e/boards/nxp/mimxrt1064_evk/mimxrt1064_evk.dts#L353-L356

So no the label used inline can be something ST-specific if that makes it more convenient.

In hope this helps shaping-up a shield:

- https://github.com/zephyrproject-rtos/zephyr/pull/89131

Commit:
drivers: video: dcmipp: add functions for external ISP functions usage

Add weak functions and their call within the dcmipp driver so that
externally provided ISP control functions can be called by the
driver at right timing in order to perform the control of the
ISP part of the DCMIPP.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

ngphibang commented on drivers/video/video_stm32_dcmipp.c:
```diff
@@ -0,0 +1,1255 @@
+		return -EINVAL;
+	}
+
+	if (fmt->pitch != fmt->width * video_bits_per_pixel(fmt->pixelformat) / BITS_PER_BYTE) {
+		LOG_ERR("Format pitch (%d) does not match with format", fmt->pitch);
+		return -EINVAL;
+	}
+
+	if ((fmt->pitch * fmt->height) > CONFIG_VIDEO_BUFFER_POOL_SZ_MAX) {
+		LOG_ERR("Frame too large for VIDEO_BUFFER_POOL");
+		return -EINVAL;
+	}
```
checking against a format based on another utility function e.g., `video_bits_per_pixel()` seems not robust as the format can be valid but may not present in the utility function, for example, I can define my own bbp in my driver:

```
static const struct mtk_camsv_format_info mtk_camsv_format_info[] = {
	{
		.fourcc = V4L2_PIX_FMT_MTISP_SBGGR12,
		.code = MEDIA_BUS_FMT_SBGGR12_1X12,
		.packed = true,
		.bpp = 12,
	}, {
		.fourcc = V4L2_PIX_FMT_MTISP_SGBRG12,
		.code = MEDIA_BUS_FMT_SGBRG12_1X12,
		.packed = true,
		.bpp = 12,
	}, {
```
Moreover, this check will be done in the sensor driver, (ii) checking against the heap size will be done when allocating buffers as it is another aspect, I think.

FYI, in https://github.com/zephyrproject-rtos/zephyr/pull/87366/commits/aa9607ed6de30c042e16e5ebc91b17d40a8525ba, format pitch is decided by the bridge driver rather by the application, so here and in get_fmt(), we can determine it rather than checking it.

avolmat-st commented on drivers/video/video_stm32_dcmipp.c:
```diff
@@ -0,0 +1,1255 @@
+	return -EINVAL;
+}
+
+static int stm32_dcmipp_set_fmt(const struct device *dev, enum video_endpoint_id ep,
+				struct video_format *fmt)
+{
```
In this first version of the serie I pushed, I kept it simple, decoupling the sensor fmt from the DCMIPP pipes output format in fact. For that purpose I'd introduced dedicated KConfig entries for that to keep the driver simple for now (at least in this first branch push).

avolmat-st commented on drivers/video/video_stm32_dcmipp.c:
This is a kind of shotcut I used since, when using the pixel pipes (aka the ones that can perform format conversion), it can be difficult to figure out which format should be used at the sensor level.
Having such define makes the driver complexity much less and thus lower the code size, considering that this mainly target embedded systems in which the system / use-case is quite defined, this might be a not too bad solution.

josuah replied:
I understand better. Making these choices in the driver would be like an application policy. The shortcut also fits well with "no policy in the kernel" but... that was for a different kernel!

Maybe some day an application library can help with these decisions... a sort of [PipelineHandler](https://wiki.st.com/stm32mpu/wiki/How_to_use_libcamera) to simplify the drivers. : ) In which case maybe [these lines](https://github.com/zephyrproject-rtos/zephyr/pull/89407/commits/5837c1efb0e0979500721a662085114053d84890#diff-41261b8df72847d560971c21aec3e0431912df38e5ed5a791a1b9633ac38cbc0R738-R741) would be removed so that the application can configure it to match, and the DCMIPP would *read*, not *write* the sensor format, to query what the user chose at runtime.

With *scalers* in the pipeline, this starts to become more interesting!
Thank you for explaining!

josuah replied:
I understand better. Making these choices in the driver would be like an application policy. The shortcut also fits well with "no policy in the kernel" but... that was for a different kernel!

Maybe some day an application library can help with these decisions... a sort of [PipelineHandler](https://wiki.st.com/stm32mpu/wiki/How_to_use_libcamera) to simplify the drivers. : ) In which case maybe [these lines](https://github.com/zephyrproject-rtos/zephyr/pull/89407/commits/5837c1efb0e0979500721a662085114053d84890#diff-41261b8df72847d560971c21aec3e0431912df38e5ed5a791a1b9633ac38cbc0R738-R741) would be removed so that the application can configure it to match, and the DCMIPP would *read*, not *write* the sensor format, to query what the user chose at runtime.

With *scalers* in the pipeline, this starts to become more interesting!
Thank you for explaining!

avolmat-st commented on drivers/video/video_stm32_dcmipp.c:
Replaced by DT_NODE_CHILD_IDX to get the index of the child

avolmat-st commented on drivers/video/video_stm32_dcmipp.c:
```diff
@@ -0,0 +1,1255 @@
+	if (Pipe == DCMIPP_PIPE0) {
+		ret = HAL_DCMIPP_PIPE_GetDataCounter(hdcmipp, Pipe, &bytesused);
+		if (ret) {
+			LOG_ERR("Failed to read counter");
```
This function returning error would most probably mean that the frame is somehow corrupted. In this situation I set the bytesused to 0 now. I will look at the POLL video API, I have never checked it yet.

avolmat-st commented:
Rebased now on top of HEAD.

DCMIPP driver now supports both CSI & Parallel input and has been tested on STM32N6 (in CSI version) and STM32MP13 (single pipe / parallel input). It should be also possible to use it on stm32h7rs however I haven't tested it on this platform.

avolmat-st commented:
Small fix to correct compliance check.

josuah commented on drivers/video/video_stm32_dcmipp.h:
[discussion]
If this is an internal header for implementing other drivers, this is the right place.
If this is a custom video API at the destination of the user, then it might be possible to put it in `<zephyr/drivers/video/stm32_dcmipp.h>` like done for other areas... A bit like a custom IOCTL in a way.

josuah replied:
Side note: if interested about an API between hardware (providing image statistics) and applications (implementing IPAs to control the ISP correction levels), then this might be of interest:
- https://github.com/zephyrproject-rtos/zephyr/pull/85811
- https://github.com/zephyrproject-rtos/zephyr/pull/87008 (not good as implements application/libraries in drivers, possible replacement: https://libmpix.org/)

avolmat-st replied:
moves to include/zephyr/video/stm32_dcmipp.h

josuah commented on boards/st/stm32n6570_dk/stm32n6570_dk_common.dtsi:
I think I lack the experience to tell what is the best, but it still looks correct.

josuah reviewed (dismissed):
Only minor non-blocking details found. LGTM!

I have the impression of taking a class on how to write a mini-V4L2 driver! :D

avolmat-st commented on drivers/video/Kconfig.stm32_dcmipp:
```diff
@@ -0,0 +1,36 @@
+
+if VIDEO_STM32_DCMIPP
+
+config VIDEO_STM32_DCMIPP_SENSOR_WIDTH
+	int "Width of the sensor frame"
+	default 2592
+	help
+	  Width of the sensor video frame.
+
+config VIDEO_STM32_DCMIPP_SENSOR_HEIGHT
+	int "Height of the sensor frame"
+	default 1944
+	help
+	  Height of the sensor video frame.
```
Indeed, good point. I'll change that. Do we agree to for the time being use that method (setting the sensor width/height/format via the Kconfig) to keep the driver simple for the time being ?

avolmat-st commented on drivers/video/video_stm32_dcmipp.c:
```diff
@@ -0,0 +1,1322 @@
+	}
+
+	pipe->active->timestamp = k_uptime_get_32();
+	pipe->active->line_offset = 0;
```
Yes that could be possible, however not yet in this version so I think it's better to stick to line_offset = 0 for now.

josuah replied:
Introducing advanced features in other PRs sounds more robust, I agree.

avolmat-st commented on drivers/video/video_stm32_dcmipp.c:
That's only 1 buffer really. The DCMIPP will update the buffer pointer every VSYNC, so here there are two pointers, one indicating the buffer currently accessed by the DCMIPP (active) and the one that will be taken on the next VSYNC (next). One a VSYNC, if next is NULL, we know that we can't have the DCMIPP use another buffer so the driver ask to the DCMIPP to stop itself on the next VSYNC.

Btw that's the way I'd like to also update the DCMI driver for, but for that we'll have to move to a lower DMA API I think since the DCMI_DMA api currently used by the DCMI driver doesn't allow such flexibility.

josuah replied:
Thank you interesting! That'd be great to allow DCMI use only 1 buffer too.

Commit:
shields: st_b_cams_imx_mb1854: add stm32n6 specifc confs

Add configuration for the STM32N6 DCMIPP driver which currently
requires to have the sensor pixel format and resolutions set via
KConfig

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

avolmat-st commented on drivers/video/video_stm32_dcmipp.c:
```diff
@@ -0,0 +1,1322 @@
+		   output_colorspace == VIDEO_COLORSPACE_RGB) {
+		/* Need to perform YUV to RGB conversion */
+		cfg = &stm32_dcmipp_yuv_to_rgb;
+	}
```
I moved the Color conversion disable within the else section since the else would be reached in such case.

ngphibang reviewed (commented):
LGTM. Thank you for adding this first "complex" driver !

ngphibang reviewed (dismissed):
As in https://github.com/zephyrproject-rtos/zephyr/pull/90919#pullrequestreview-2895588005, CI failure is not due to the PR change.

josuah commented on drivers/video/video_stm32_dcmipp.c:
```diff
@@ -0,0 +1,1322 @@
+#endif
+#if defined(DCMIPP_PIPE1) && defined(DCMIPP_PIPE2)
+#define STM32_DCMIPP_HAS_PIXEL_PIPES
+#endif
```
That looks ideal!

erwango reviewed (approved):
Nice, LGTM.

It would be welcomed to document feature support at board level (fact that feature is available  with the shield), but I won't block at this stage. Please add it in a follow up PR though.
